kvbc/tools/db_restore/db_restore.cpp

```diff
@@ -0,0 +1,122 @@
+      "rocksDB directory path");
+  cli_options_.add_options()("help,h", "produce help message");
+  cli_options_.add(integrity_check_.getOptions()).add(cli_mandatory_options_);
+  // clang-format on
+}
+void DBRestore::parseCLIArgs(int argc, char** argv) {
+  auto usage = [this]() { LOG_ERROR(logger_, cli_options_); };
+
+  // parse arguments
+  po::store(po::parse_command_line(argc, argv, cli_options_), var_map_);
+
+  // check arguments
+  try {
+    po::notify(var_map_);
+  } catch (const std::exception& e) {
+    usage();
+    throw;
+  }
+  if (var_map_.count("help")) {
+    usage();
+    std::exit(1);
+  }
+}
+
+void DBRestore::initRocksDB(const fs::path& rocksdb_path) {
+  LOG_DEBUG(logger_, rocksdb_path);
+  rocksdb_dataset_ = std::make_unique<v2MerkleTree::RocksDBStorageFactory>(rocksdb_path)->newDatabaseSet();
+  const auto linkStChain = true;
+  auto kvbc_categories = std::map<std::string, categorization::CATEGORY_TYPE>{
+      {categorization::kExecutionProvableCategory, categorization::CATEGORY_TYPE::block_merkle},
+      {categorization::kExecutionPrivateCategory, categorization::CATEGORY_TYPE::versioned_kv},
+      {categorization::kExecutionEventsCategory, categorization::CATEGORY_TYPE::immutable},
+      {categorization::kRequestsRecord, categorization::CATEGORY_TYPE::immutable},
+      {categorization::kExecutionEventGroupDataCategory, categorization::CATEGORY_TYPE::immutable},
+      {categorization::kExecutionEventGroupTagCategory, categorization::CATEGORY_TYPE::immutable},
+      {categorization::kExecutionEventGroupLatestCategory, categorization::CATEGORY_TYPE::versioned_kv},
+      {categorization::kConcordInternalCategoryId, categorization::CATEGORY_TYPE::versioned_kv},
+      {categorization::kConcordReconfigurationCategoryId, categorization::CATEGORY_TYPE::versioned_kv}};
+
+  kv_blockchain_ = std::make_unique<categorization::KeyValueBlockchain>(
+      storage::rocksdb::NativeClient::fromIDBClient(rocksdb_dataset_.dataDBClient), linkStChain, kvbc_categories);
+}
+/**
+ * Optimistic approach for a single pass.
+ * Retrieve blocks from object store in a natural order and add them to rocksdb while comparing their digests.
+ * At the end compare the last block's digest with a one in latest checkpoint descriptor.
+ */
+void DBRestore::restore() {
+  auto [last_os_block_id, last_os_block_digest] = integrity_check_.getLatestsCheckpointDescriptor();
+  auto last_reachable_rocks_blockid = kv_blockchain_->getLastReachableBlockId();
+  Digest last_reachable_rocksdb_block_digest;
+  if (last_reachable_rocks_blockid == 0) {
+    LOG_WARN(logger_, "rocksdb is empty, will restore the whole blockchain");
+  } else {
+    auto buffer =
+        categorization::RawBlock::serialize(kv_blockchain_->getRawBlock(last_reachable_rocks_blockid).value());
+    auto last_reachable_rocks_block = Sliver::copy(reinterpret_cast<const char*>(buffer.data()), buffer.size());
```

Comment:
Why is the copy needed?

Reply:
because Sliver owns the buffer and there will be a double free otherwise

Reply:
I meant whether the copy is necessary for the functional correctness of the implementation, 
because it is a rather large copy.

Reply:
fixed to string_view